internal/metrics/definition.go

```diff
@@ -0,0 +1,146 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package metrics
+
+import (
+	"errors"
+	"sync"
+
+	"go.opentelemetry.io/otel"
+	api "go.opentelemetry.io/otel/metric"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	"github.com/envoyproxy/gateway/api/v1alpha1"
+	log "github.com/envoyproxy/gateway/internal/logging"
+)
+
+var (
+	meter = func() api.Meter {
+		return otel.GetMeterProvider().Meter("envoy-gateway")
+	}
+
+	monitoringLogger = log.DefaultLogger(v1alpha1.LogLevelInfo).WithName("metrics")
+)
+
+func init() {
+	otel.SetLogger(monitoringLogger.Logger)
+}
+
+// MetricType is the type of a metric.
+type MetricType string
+
+// Metric type supports:
+// * Counter: A Counter is a simple metric that only goes up (increments).
+//
+// * Gauge: A Gauge is a metric that represent
+// a single numerical value that can arbitrarily go up and down.
+//
+// * Histogram: A Histogram samples observations and counts them in configurable buckets.
+// It also provides a sum of all observed values.
+// It's used to visualize the statistical distribution of these observations.
+
+const (
+	CounterType   MetricType = "Counter"
+	GaugeType     MetricType = "Gauge"
+	HistogramType MetricType = "Histogram"
+)
+
+// A Metric collects numerical observations.
+type Metric interface {
```

Comment:
is these something from istio/istio/pkg/monitoring?

Reply:
The interface abstraction refers the metric design in istio, did not figure out more, top level metric abstraction should be similar, but with different implementations and registers/exporters.

Reply:
I think there's some lib like this, at least, you should not copy the code or add reference.

Reply:
not 100%, but maybe https://github.com/tetratelabs/telemetry can help.

Reply:
https://github.com/tetratelabs/telemetry/blob/master/metric.go the metric difinitions are pretty similar to istio/istio/pkg/monitoring too, as I said, the top level interface abstraction of metric should be similar.

Reply:
Or are you proposing directly use `github.com/tetratelabs/telemetry` as the Interfaces, and we do implementations like https://github.com/tetratelabs/telemetry-opencensus? But this will make us not that easy to maintain the metrics library if we want to add/change more interfaces

Reply:
yes, we should not copy everything from istio or do something similar.

Reply:
Current version is not a copy from istio, different in both interfaces and implementations. I have refactored a couple of times, and tried directly to use tetrate/telemtry or istio/pkg/monitoring, both ways are not ideal and not suitable for eg, which spent me a lot of time on it. I believe current version is the ideal design for eg comp metrics obs.